### PR TITLE
Recognize __() and _() methods as camel case (Issue #839)

### LINKED ISSUE #839: "__()" method is marked as not camel caps

### PR DESCRIPTION
**Pull request by tim-bezhashvyly**
Closes #839.

### CODE REVIEW

File: CodeSniffer.php [hunk 1]
```diff
@@ -1933,6 +1933,10 @@ public static function isCamelCaps(
         $public=true,
         $strict=true
     ) {
+        if ('' === $string) {
```
Comment (aik099) [inline CodeSniffer.php]: Please do not use Yoda-style IF conditions.

Comment (tim-bezhashvyly) [inline CodeSniffer.php]: Why?

**Comment by aik099** [comment]
I think that relevant sniff should be fixed instead. Technically there can't be empty method name and it's the sniff that produces it. This means that sniff does wrong thing.

File: CodeSniffer.php (hunk 1 above)
Comment (aik099) [inline CodeSniffer.php]: Because this is coding standard used in this library. I know that Yoda-style is highly appreciated in Symfony components though. Not really sure why and how it all started.

**Comment by tim-bezhashvyly** [comment]
I thought the same first but here we are not talking about a sniff but about `isCamelCaps` method.

File: CodeSniffer.php (hunk 1 above)
Comment (tim-bezhashvyly) [inline CodeSniffer.php]: I'm not a Symfony developer. My intention of using a Yoda notations is to avoid `if ($string = '') { }`. I agree that Yoda notations are highly misused as in cases like `if (1 < $foo) { }` or `if ('' === trim($string))` it make no sense, still in this particular example Yoda notation can prevent a stupid typo which is quite hard to debug.

Saying all this I'm don't want to break your coding standards if they are strictly prohibiting Yoda notations. In this case please let me know and I will flip it.

**Comment by aik099** [comment]
> I thought the same first but here we are not talking about a sniff but about isCamelCaps method.

It's place where you decided to make a fix, but I'm saying that you should fix source of the problem (the sniff that calls `isCamelCaps('')` method, not adding obscure condition saying empty method name is camel caps ;)

File: CodeSniffer.php (hunk 1 above)
Comment (aik099) [inline CodeSniffer.php]: Then do it ;), because I already requested that change few inline comments above, but we're just chatting about the Yoda conditions.

**Comment by tim-bezhashvyly** [comment]
I definitely can not agree with you here.

`isCamelCaps('')` is a valid use case. I want to know if the string is camel caps or not. I don't need to pre-filter strings before. This will create exceptions special code flow and this must be avoided.

The source of problem it not a test but the `isCamelCaps` method which can not tell me correctly if the string is camel caps or not.

File: CodeSniffer.php (hunk 1 above)
Comment (tim-bezhashvyly) [inline CodeSniffer.php]: Flipped.

**Comment by aik099** [comment]
Let's wait for @gsherwood decision on this then.

**Comment by gsherwood** [comment]
An empty string should probably throw an exception. It's the very definition of an undefined result.

I would then be checking the empty string in the sniffs themselves, be they function, method or class sniffs (or anything else that might use the camelcase check).

**Comment by aik099** [comment]
@gsherwood , thanks for confirming my initial assertion of changes that needs to be made in this PR.

**Comment by tim-bezhashvyly** [comment]
Reimplemented on a sniff side.

File: CodeSniffer/Standards/PSR1/Sniffs/Methods/CamelCapsMethodNameSniff.php [hunk 2]
```diff
@@ -34,16 +34,6 @@ class PSR1_Sniffs_Methods_CamelCapsMethodNameSniff extends Generic_Sniffs_Naming
 
 
     /**
-     * Constructs a PSR1_Sniffs_Methods_CamelCapsMethodNameSniff.
-     */
-    public function __construct()
```
Comment (aik099) [inline CodeSniffer/Standards/PSR1/Sniffs/Methods/CamelCapsMethodNameSniff.php]: Why have you removed this method?

File: CodeSniffer/Standards/PSR1/Tests/Methods/CamelCapsMethodNameUnitTest.inc [hunk 3]
```diff
@@ -61,5 +61,6 @@ class Foo extends \SoapClient
     ) {
         // body
     }
+
+    function __() {}
 }
-?>
```
Comment (aik099) [inline CodeSniffer/Standards/PSR1/Tests/Methods/CamelCapsMethodNameUnitTest.inc]: This is an unrelated change. Please remove here and in other 2 files.

If you want to remove `?>` in whole test suite or library, then please do this in dedicated PR.

File: CodeSniffer/Standards/PSR1/Sniffs/Methods/CamelCapsMethodNameSniff.php (hunk 2 above)
Comment (tim-bezhashvyly) [inline CodeSniffer/Standards/PSR1/Sniffs/Methods/CamelCapsMethodNameSniff.php]: Because it is pointless. The constructor of a parent is not taking any arguments.

Comment (aik099) [inline CodeSniffer/Standards/PSR1/Sniffs/Methods/CamelCapsMethodNameSniff.php]: And this is not change related to PR topic. Please move it to separate PR so that best solution can be discussed separately.

I think, that better approach would be changing parent class constructor to accept 3 given arguments and defaults to currently used values (in parent class).

Comment (tim-bezhashvyly) [inline CodeSniffer/Standards/PSR1/Sniffs/Methods/CamelCapsMethodNameSniff.php]: Restored.

**Comment by aik099** [comment]
Great, now this PR is on topic and after squashing can be merged.

P.S.
Don't forgot to submit 2 other PR's with changes that were not related to this PR.

**Comment by tim-bezhashvyly** [comment]
I won't. Frankly I gave up fighting with you guys. Only interested in this particular PR as it breaks my build.

**Comment by aik099** [comment]
> Frankly I gave up fighting with you guys.

Sorry if that looked that way. I understand, that every person has his/her point of view, but so do the library maintainers. Since this is probably your 1st PR to the library you might not be aware of all the specifics and we gladly informed you about them. This turned out not what you've expected, but that's life.

If we'd be merging every PR without checking it first, then this could ended up in a mess and that's not good in any case.

Anyway thanks for taking your time to make a contribution and don't forget to squash commits in your PR.

P.S.
The "On topic PR" is rule, which applies to most projects and violating it most likely would result in PR not being merged. Following it can speed up further PR merging on this or any other project.

**Comment by tim-bezhashvyly** [comment]
This is not that's I'm disappointed. More like our vision on clean and readable code dramatically differs.

**Comment by aik099** [comment]
> More like our vision on clean and readable code dramatically differs.

That's not mine vision, but vision of this library. If you're interested in my vision specifically :wink: , then you might look at projects I've created, like:

* https://github.com/qa-tools/qa-tools
* https://github.com/console-helpers/svn-buddy
* https://github.com/minkphp/phpunit-mink

**Comment by tim-bezhashvyly** [comment]
Took a quick look at QA Tools. Still lots of things I dislike such as protected methods and properties, non-descriptinve names which spawn unneeded comments which makes code less readable. In addition to it there are some smelly practices as overusage of inheritance, usage of constructors for instantiation and general SOLID violations.

Sorry for spitting it out but I'm used to tell what I think.

**Comment by gsherwood** [comment]
I'm currently on holidays, but jumped in to make sure my position on empty string handling was known as I wont be back until next week. @tim-bezhashvyly Since then, this PR has degraded to the point where it appears you are not happy. I wasn't aware there was any fight going on, but I have not been following PRs as closely as normal. I apologise if you feel like your contribution hasn't been handled appropriately.

I've had a look at the new commits and they look perfectly fine to me. It's how I would have fixed things myself.

@tim-bezhashvyly If this method of fixing the issue is not acceptable to you, I'm perfectly happy to make the commits myself so you are not associated with them. Please let me know if you'd prefer that or are happy for me to merge the PR as is (I'll do this when I'm back at work next week).

**Comment by tim-bezhashvyly** [comment]
@gsherwood: Current state of this PR solves my issue. This suits me fine. Thank you.

**Comment by gsherwood** [comment]
> @gsherwood: Current state of this PR solves my issue. This suits me fine. Thank you.

Thanks for getting back to me. And thanks for reporting the issue. I'm aware that `__()` methods are commonly used for translation, and of course the `_()` alias for gettext itself, so I'm glad it has been found and will be fixed in the next release.

**Comment by aik099** [comment]
> Took a quick look at QA Tools. Still lots of things I dislike such as protected methods and properties, non-descriptinve names which spawn unneeded comments which makes code less readable. In addition to it there are some smelly practices as overusage of inheritance, usage of constructors for instantiation and general SOLID violations.
> 
> Sorry for spitting it out but I'm used to tell what I think.

If you have time, please report any things you found disturbing as issues so that it can be discussed with other library contributors as well. Don't create any PR's yet until we've agreed on any changes to be made to avoid spending too much time on something that might not be merged (actually I've written that in contribution docs as well).

**Comment by tim-bezhashvyly** [comment]
It's there a page about the 3.0 branch, its purpose, goals and status?

**Comment by gsherwood** [comment]
> It's there a page about the 3.0 branch, its purpose, goals and status?

There isn't.

But in summary, the 3.0 branch is where I am attempting to restructure the codebase to allow me to more easily maintain the project and add features that are hard to do in the current version (caching and parallel processing were the main ones I added in there).

The goal is to do this without breaking backwards compatibility too much, but that will happen for anyone who has written their own sniffs. I will have an upgrade guide when the 3.0 version is ready for an alpha release.

My current rough changelog notes are here (top section only, until the blank line): https://github.com/squizlabs/PHP_CodeSniffer/blob/3.0/package.xml#L29
